### DISCUSSION TITLE
please elaborate on FAQ entry (No .env file found)

### DISCUSSION THREAD

**Discussion by tessus** [dani-garcia/vaultwarden#3090] [Q&A] [answered]
I came across the FAQ entry below, but it doesn't make sense to me.

>Why does Vaultwarden say [INFO] No .env file found. even though I provided one?
>
>While launching, Vaultwarden checks for the existence of a file called .env (if not changed via the environment variable ENV_FILE) in the current working directory of the process. If you have not provided this file Vaultwarden will simply inform you that it has not found it. This file is not related to the env file you have provided to docker which loads in the environment variables on container creation or the EnvironmentFile used within a [systemd .service](https://github.com/dani-garcia/vaultwarden/wiki/Setup-as-a-systemd-service).

So... Why do I need a `.env` file in the first place? All the info is either in the file provided  to docker or in the env file specified in the systemd service.
It makes no sense to me that vaultwarden even checks for that file.

Can someone please explain to me the logic and flow?

e.g. what happens if there's the env file in systemd and a `.env` file next to the vaultwarden binary? Are they merged?

IMO vaultwarden should only check for that file if nothing is provided. And/or vaultwarden should print which env file is actually used.

P.S.: The fact that every time I start/restart vw and `systemctl status` shows me `[INFO] No .env file found.` adds cognitive load, because this info is totally useless, because my vw is configured correctly, there is no need to mention that an irrelevant file has not been found.

**Comment by stefan0xC** [comment]
> So... Why do I need a `.env` file in the first place? All the info is either in the file provided to docker or in the env file specified in the systemd service.
> It makes no sense to me that vaultwarden even checks for that file.
>
> Can someone please explain to me the logic and flow?

You can run vaultwarden without docker or as a systemd unit.

> e.g. what happens if there's the env file in systemd and a .env file next to the vaultwarden binary? Are they merged?

The loaded environment variables  are merged with the env file (while preserving already loaded environment variables) which is done through the use of the `dotenvy` crate. https://github.com/dani-garcia/vaultwarden/blob/10dadfca068ed449fcd4a74b70ae2cd83990d3d4/src/config.rs#L63

> IMO vaultwarden should only check for that file if nothing is provided. And/or vaultwarden should print which env file is actually used.

That would not work. As explained above the environment files for docker/systemd are completely unrelated to the `ENV_FILE`/`.env` file Vaultwarden is looking for and a process does not know how the already provided environment variables (if any) are loaded.

**Reply by tessus** [reply]
> You can run vaultwarden without docker or as a systemd unit.

I do, which is why I do not need the .env file. ;-) See further explanation in my reply below.

**Reply by stefan0xC** [reply]
I meant neither.

**Comment by BlackDex** [comment]
That file is used mostly by standalone/containerless systems.
So, if you build the binary your self and then execute it and have a `.env` file in the current working directory, it will use that.

Docker-compose does the exact same thing, but then you can use those variables in your docker-compose.yaml file. Same for systemd, it will use the configured file to populate the environment with those variables.

Vaultwarden will always try to find an read that file and use those values. Now from the top of my head i think the value from the fire overrule the custom set environment variables, but I'm not sure.

Also, don't forget the `config.json` file which, if that exist will overrule everything else.

**Reply by tessus** [reply]
> That file is used mostly by standalone/containerless systems.

I use and create my own RPMs for Fedora and I certainly do not use a `.env` file in the working directory. e.g. I use `/etc/vaultwarden/vaultwarden.cfg`.

It's just very confusing to be reminded that a `.env` file was not found, because:
- I don't need that .env file
- this message has a slightly negative connotation

So in my case vaultarden reads the config from `/etc/vaultwarden/vaultwarden.cfg`. There's no need to tell me that a .env file was not found. However, it would make more sense to show in the log that the config is read from `/etc/vaultwarden/vaultwarden.cfg`.

P.S.:

> Also, don't forget the config.json file which, if that exist will overrule everything else.

Yep, this file is only created when one uses the admin page to change the config. e.g. on my system the file does not exist. But the log does not say that config.json was not found. It only tells me that the .env file was not found, which I am not using either.
It's inconsistent. 

IMO I only need a message that the config file was not found, when I specify a config file and it is not there. ;-)

**Reply by BlackDex** [reply]
> However, it would make more sense to show in the log that the config is read from `/etc/vaultwarden/vaultwarden.cfg`.

If you would configure `ENV_FILE` to point to that file (if it is a .env file formatted file), then you will not see that message about not finding the .env file.

**Reply by BlackDex** [reply]
Though, that said, it's something on my list to take a look at for a while, maybe i can check it out today.

**Reply by tessus** [reply]
In my systemd file I set `EnvironmentFile=/etc/vaultwarden/vaultwarden.cfg`.

So I should only get a message in the log, if that file didn't exist.

**Reply by tessus** [reply]
It's not something super important just a bit confusing. As mentioned in my [P.S.](https://github.com/dani-garcia/vaultwarden/discussions/3090#discussioncomment-4569849), I don't get a message because of a missing config.json either.

**Reply by BlackDex** [reply]
No, you will never received a message from Vaultwarden regarding the `EnvironmentFile=`, because Vaultwarden doesn't know of it's existence. Systemd populates the environment with that file, it doesn't provide that file to Vaultwarden.

Also, you will never received a message regarding a missing `config.json`, but you will receive a message that it is using that and that variables are overrules by it.

**Reply by tessus** [reply]
Maybe the solution is to not set `EnvironmentFile=`, but `Environment=ENV_FILE=/etc/vaultwarden/vaultwarden.cfg`

I actually like the concept of `EnvironmentFile` better, but if that always yields that annoying message, I have to change it.

**Reply by BlackDex** [reply]
Or wait for some improvements, or, just ignore it. It's just an info message after all

**Reply by stefan0xC** [reply]
> Maybe the solution is to not set `EnvironmentFile=`, but `Environment=ENV_FILE=/etc/vaultwarden/vaultwarden.cfg`
> 
> I actually like the concept of `EnvironmentFile` better, but if that always yields that annoying message, I have to change it.

You could do both to get rid of the log message.

**Comment by BlackDex** [comment, marked as answer]
@tessus See #3092, where i have changed the way the messages are being shown.
Hopefully this resolves your unwanted message.

But more i hope that this will not cause other users to ask questions regarding the missing `.env` file.

**Reply by tessus** [reply]
Thanks for this. I believe the flow is much clearer now.

A few months after the merge (so that most people have upgraded to the new release) you can probably even delete the FAQ entry. ;-)
